velox/functions/prestosql/json/JsonStringUtil.cpp

```diff
@@ -0,0 +1,175 @@
+#include "velox/functions/prestosql/json/JsonStringUtil.h"
+
+using namespace facebook::velox::functions;
+
+namespace facebook::velox {
+namespace {
+
+FOLLY_ALWAYS_INLINE char hexDigit(uint8_t c) {
+  VELOX_DCHECK_LT(c, 16);
+  return c < 10 ? c + '0' : c - 10 + 'A';
+}
+
+FOLLY_ALWAYS_INLINE void writeHex(char16_t value, char*& out) {
+  value = folly::Endian::little(value);
+  *out++ = '\\';
+  *out++ = 'u';
+  *out++ = hexDigit((value >> 12) & 0x0F);
+  *out++ = hexDigit((value >> 8) & 0x0F);
+  *out++ = hexDigit((value >> 4) & 0x0F);
+  *out++ = hexDigit(value & 0x0F);
+}
+
+std::array<int8_t, 128> getAsciiEscapes() {
+  std::array<int8_t, 128> escapes;
+  std::fill(escapes.data(), escapes.data() + 32, -1);
+  escapes['"'] = '"';
+  escapes['\\'] = '\\';
+  escapes['\b'] = 'b';
+  escapes['\t'] = 't';
+  escapes['\n'] = 'n';
+  escapes['\f'] = 'f';
+  escapes['\r'] = 'r';
+  return escapes;
+}
+static const std::array<int8_t, 128> asciiEscapes = getAsciiEscapes();
+
+FOLLY_ALWAYS_INLINE void encodeAscii(int8_t value, char*& out) {
+  int8_t escapeCode = asciiEscapes[value];
+  if (escapeCode == 0) {
+    *out++ = char(value);
+  } else if (escapeCode > 0) {
+    *out++ = '\\';
+    *out++ = char(escapeCode);
+  } else {
+    writeHex(value, out);
+  }
+}
+
+std::array<uint8_t, 128> getEncodedAsciiSizes() {
+  std::array<uint8_t, 128> sizes;
+  std::fill(sizes.data(), sizes.data() + 32, 6);
+  std::fill(sizes.data() + 32, sizes.data() + 128, 1);
+  for (int8_t c : {'"', '\\', '\b', '\t', '\n', '\f', '\r'}) {
+    sizes[c] = 2;
+  }
+  return sizes;
+}
+static const std::array<uint8_t, 128> encodedAsciiSizes =
+    getEncodedAsciiSizes();
+
+// Encode `codePoint` value into one or two UTF-16 code units. Write each code
+// unit as prefixed hexadecimals of 6 chars.
+FOLLY_ALWAYS_INLINE void encodeUtf16Hex(char32_t codePoint, char*& out) {
+  VELOX_DCHECK(codePoint <= 0x10FFFFu);
+  // Two 16-bit code units are needed.
+  if (codePoint >= 0x10000u) {
+    writeHex(
+        static_cast<char16_t>(
+            0xD800u + (((codePoint - 0x10000u) >> 10) & 0x3FFu)),
+        out);
+    writeHex(
+        static_cast<char16_t>(0xDC00u + ((codePoint - 0x10000u) & 0x3FFu)),
+        out);
+    return;
+  }
+  // One 16-bit code unit is needed.
+  writeHex(static_cast<char16_t>(codePoint), out);
+}
+
+} // namespace
+
+void testingEncodeUtf16Hex(char32_t codePoint, char*& out) {
+  encodeUtf16Hex(codePoint, out);
+}
+
+void escapeString(const char* input, size_t length, char* output) {
+  char* pos = output;
+
+  auto* start = reinterpret_cast<const unsigned char*>(input);
+  auto* end = reinterpret_cast<const unsigned char*>(input + length);
+  while (start < end) {
+    int count = validateAndGetNextUtf8Length(start, end);
+    switch (count) {
+      case 1: {
+        encodeAscii(int8_t(*start), pos);
+        start++;
+        continue;
+      }
+      case 2:
+      case 3: {
+        memcpy(pos, reinterpret_cast<const char*>(start), count);
```

Comment:
You need to use the literals `2` and `3`, variable like `count` cannot be optimized by compiler and result in very slow code: https://godbolt.org/z/n8rajeYcc

Reply:
Thank you, @Yuhta , for writing and showing the compiled code for me!